### DISCUSSION TITLE
What is the best way to investigate pyright performance.

### DISCUSSION THREAD

**Discussion by parched** [microsoft/pyright#4516] [Q&A] [answered]
On the main part of our project, pyright takes about 4min to run. We'd like to use it as a precommit hook but it's too slow. Currently we just run it on the files that have changed, but that only really catches the errors you'd notice while editing the files anyway.

Our code base is being progressively typed, so there's a lot of untyped code at the moment. I'd like to know if there's any low hanging fruit of untyped code that slows pyright down that I could fix. I ran it with `--stats` but that didn't provide anything insightful.
```
1456 errors, 46319 warnings, 0 informations
Completed in 278.763sec

Analysis stats
Total files parsed and bound: 1471
Total files checked: 1003

Timing stats
Find Source Files:    0.24sec
Read Source Files:    0.48sec
Tokenize:             1.11sec
Parse:                2.24sec
Resolve Imports:      1.56sec
Bind:                 2.32sec
Check:                236.49sec
Detect Cycles:        0sec
```

**Comment by erictraut** [comment, marked as answer]
Here are a few things you can try:
1. Use the `--stats` along with `--verbose`. This will print out a lot of output, but at the end it will include a list of files with analysis times (sorted from largest to smallest). This will help you localize the analysis time to a particular file or files.
2. You can try using the `--skipunannotated` option to limit type evaluation to functions whose signatures are annotated.

**Reply by parched** [reply]
Ah `--verbose`, I should've thought of that, thanks!

It's still running, but the logs seem to look like this over and over.
![image](https://user-images.githubusercontent.com/5975405/214192730-baa2dcab-9afc-40e3-b6ab-0487b754e6a3.png)
What does `Ae` and `Ns` mean?

**Reply by parched** [reply]
Hmm, the worst offender has no type errors or warnings and is not very long (506251ms).
All of the tested functions are typed to accept and return `DOFResult = numpy.typing.NDArray[np.float32]`
```
import numpy as np
import pytest

import holmes_model.lib as hm


def test_combine_results_via_direct_sum__return_summed_items():

    test_input = np.array([[1, 2],
                           [3, 4]]).reshape(1, 1, 2, 1, 1, 2)
    result_data = hm.ResultsCombining.combine_results_via_direct_sum(test_input)

    assert np.all(result_data == np.array([[[[[[4, 6]]]]]]))


def test_combine_results_via_envelope_with_opposite_sign__return_enveloped_items():
    """Test combinations of bigger, smaller, bigger (-ve), smaller (-ve)"""

    test_input = np.array([[2, 2, 2, 2, -2, -2, -2, -2],
                           [3, 1, -3, -1, 3, 1, -3, -1]]).reshape(1, 1, 2, 1, 1, 8)
    result_data = hm.ResultsCombining.combine_results_via_envelope(test_input)

    assert np.all(result_data == np.array([[[[[[3, 2, -3, 2, 3, -2, -3, -2]]]]]]))


def test_combine_results_via_envelope__with_single_set__return_opposite_set():

    test_input = np.array([2, -2]).reshape(1, 1, 1, 1, 1, 2)
    result_data = hm.ResultsCombining.combine_results_via_envelope_with_opposite_sign(test_input)

    assert np.all(result_data == np.array([[[[[2, -2], [-2, 2]]]]]))


def test_combine_results_via_envelope__with_single_set__return_opposite_enveloped_set():

    test_input = np.array([[2, 2],
                           [3, -4]]).reshape(1, 1, 2, 1, 1, 2)
    result_data = hm.ResultsCombining.combine_results_via_envelope_with_opposite_sign(test_input)

    assert np.all(result_data == np.array([[[[[3, -4], [-3, 4]]]]]))


def test_combine_results_via_absolute_sum__return_summed_abs_items():

    test_input = np.array([[-1, 2, 3],
                           [-3, 4, -8]]).reshape(1, 1, 2, 1, 1, 3)
    result_data = hm.ResultsCombining.combine_results_via_absolute_sum(test_input)

    assert np.all(result_data == np.array([[[[[[4, 6, 11]]]]]]))


def test_combine_results_via_grafted_envelope_pair__return_grafted_results():
    # NOTE the method flips the initial values because we want +X first (typically -X is in the base envelope)

    test_input = np.array([[10, -10],
                           [1, -1]]).reshape(1, 1, 2, 1, 2, 1)
    result_data = hm.ResultsCombining.combine_results_via_grafted_envelope_pair(test_input)

    assert np.all(result_data == np.array([[[[[[11], [9], [-9], [-11]]]]]]))


def test_combine_results_via_grafted_envelope_pair__with_one_set_items_in_axis4__raise_assertion_error():
    test_input = np.array([[10, -10],
                           [1, -1]]).reshape(1, 1, 2, 1, 1, 2)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_via_grafted_envelope_pair(test_input)


def test_combine_results_via_grafted_envelope_pair__with_one_set_items_in_axis2__raise_assertion_error():
    test_input = np.array([[10, -10],
                           [1, -1]]).reshape(1, 1, 1, 1, 2, 2)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_via_grafted_envelope_pair(test_input)


def test_combine_results_via_stitched_envelope__with_single_set_items_in_axis4__return_items_joined_in_axis2():

    test_input = np.array([[2, 2],
                           [-1, -2]]).reshape(1, 1, 2, 1, 1, 2)
    result_data = hm.ResultsCombining.combine_results_via_stitched_envelope(test_input)

    assert np.all(result_data == np.array([[[[[[2, 2], [-1, -2]]]]]]))


def test_combine_results_via_stitched_envelope__with_two_sets_items_in_axis4__raise_assertion_error():

    test_input = np.array([[2, 2], [-1, -2],
                           [2, 2], [-1, -2]]).reshape(1, 1, 2, 1, 2, 2)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_via_stitched_envelope(test_input)


def test_combine_results_via_no_combination__return_same_item_with_squeezed_array():

    test_input = np.array([1, 2]).reshape(1, 1, 1, 1, 1, 2)
    result_data = hm.ResultsCombining.combine_results_via_direct_sum(test_input)

    assert np.all(result_data == np.array([[[[[[1, 2]]]]]]))


def test_combine_results_via_no_combination__with_more__than_one_load_entity__raise_assertion_error():

    test_input = np.array([[1, 2],
                           [3, 4]]).reshape(1, 1, 2, 1, 1, 2)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_no_combination(test_input)


def test_combine_results_via_grafted_envelope_single__return_grafted_results():

    test_input = np.array([[10], [1]]).reshape(1, 1, 2, 1, 1, 1)
    result_data = hm.ResultsCombining.combine_results_via_grafted_envelope_single(test_input)

    assert np.all(result_data == np.array([[[[[[11], [9], [-9], [-11]]]]]]))


def test_combine_results_via_grafted_envelope_single__with_two_set_items_in_axis4__raise_assertion_error():
    test_input = np.array([[10, -10],
                           [1, -1]]).reshape(1, 1, 2, 1, 2, 1)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_via_grafted_envelope_single(test_input)


def test_combine_results_via_grafted_envelope_single__with_one_set_items_in_axis2__raise_assertion_error():
    test_input = np.array([[10], [1]]).reshape(1, 1, 1, 1, 1, 2)

    with pytest.raises(AssertionError):
        hm.ResultsCombining.combine_results_via_grafted_envelope_single(test_input)
```

**Reply by erictraut** [reply]
The analysis times aren't exact because analysis doesn't strictly occur on file boundaries. It's likely that the file in question happens to import some symbol from another file, and that symbol is expensive in terms of type analysis. I'm not familiar with `holmes_model.lib`. It's possible that's the culprit. Is it a typed library?

Do you have `useLibraryCodeForTypes` enabled? If so, you could disable that. Inferring types from library code is often very expensive. It's much better to use typed libraries and/or provide your own stubs for libraries that are untyped.

**Reply by parched** [reply]
`holmes_model` is our private project, but I'll try and extract a few files that I can share to see if it still reproduces.

The next file on the list is about 700 lines of `dict` and `list` constants containing enums, that took 31440ms.

**Reply by parched** [reply]
Actually, there is some type warnings, I just didn't notice because they took so long to show up in VS Code. @erictraut I've given you access to a repo with a few of the files, if you're interested in testing it. Looks like it something to do with all the combinations of overloads in numpy.

**Reply by erictraut** [reply]
I'm able to repro the problem. It's related to a combination of 1) a recursive protocol type (`numpy.typing._NestedSequence`) that includes overloads, 2) complex overloads in `numpy.array`, and 3) deeply nested list arguments in your code like `[[[[[[4, 6]]]]]]`. Each of these affects the analysis time in a multiplicative manner.

I've found one optimization (related to protocol matching when overloads are present) that reduces the analysis time from about 43 seconds to about 4 seconds on my machine. That's a big improvement, but 4 seconds is still a long time. I'll continue to look for additional improvements.

**Reply by parched** [reply]
Great, thank you Eric. I've also just added an excerpt (constants.py) from the second worst file in our repo. It pattern matches on an enum with 200+ cases. Having found this horrible enum, I'm going to refactor it, but it's probably worth investigating the performance as I guess some may legitimately have an enum that big.

**Reply by erictraut** [reply]
This results in a union with over 260 items, and narrowing it in a match statement requires quite a bit of work — effectively an n^2 operation every time one enum is removed from the union during the narrowing process. I've fixed the perf issue by adding a map to eliminate the n^2 behavior when a union contains a large number of enum literals. This takes the analysis time of that file from more than 6000ms to under 100ms.

**Reply by erictraut** [reply]
@parched, I just published pyright 1.1.291. I'm interested to see how it performs on your full code base. Could you post the "--stats" output?

**Reply by parched** [reply]
It's much better, thanks!
```
1521 errors, 43880 warnings, 0 informations 
Completed in 87.125sec

Analysis stats
Total files parsed and bound: 1470
Total files checked: 1003

Timing stats
Find Source Files:    0.2sec
Read Source Files:    0.42sec
Tokenize:             0.84sec
Parse:                1.98sec
Resolve Imports:      1.59sec
Bind:                 2.08sec
Check:                56.9sec
Detect Cycles:        0sec
```

I'll dig into the verbose output tomorrow, to see if there's any other notable slow spots.
